proposals/NNNN-extensible-enums.md

```diff
@@ -0,0 +1,185 @@
+non-exhaustive list of problems this can cause:
+
+- Using enumerations to represent `Error`s is inadvisable, as if new errors need
+  to be introduced they cannot be added to existing enumerations. This leads to
+  a proliferation of `Error` enumerations. "Fake" enumerations can be made using
+  `struct`s and `static let`s, but these do not work with the nice `Error`
+  pattern-match logic in catch blocks, requiring type casts.
+- Using an enumeration to refer to a group of possible ideas without entirely
+  exhaustively evaluating the set is potentially dangerous, requiring a
+  deprecate-and-replace if any new elements appear.
+- Using an enumeration to represent any concept that is inherently extensible is
+  tricky. For example, `SwiftNIO` uses an enumeration to represent HTTP status
+  codes. If new status codes are added, SwiftNIO needs to either mint new
+  enumerations and do a deprecate-and-replace, or it needs to force these new
+  status codes through the .custom enum case.
+
+This proposal plans to address these limitations on enumerations in
+non-resilient Swift.
+
+## Proposed solution
+
+We propose to introduce a new language feature `ExtensibleEnums` that aligns the
+behaviour of enumerations in both language dialetcs. This will make enumerations
+in packages a safe default and leave maintainers the choice of extending them
+later on.
+
+In modules with the language feature enabled, developers can use the exisiting
+`@frozen` attribute to mark an enumeration as non-extensible. Allowing consumers
+of the module to exhaustively switch over the cases. This makes commiting to the
+API of an enum an active choice for developers.
+
+Modules consuming other modules with the language feature enabled will be forced
+to add an `@unknown default:` case to any switch state for enumerations that are
+not marked with `@frozen`.
+
+Since enabling a language feature applies to the whole module at once we also
+propose adding a new attribute `@extensible` analogous to `@frozen`. This
+attribute allows developers to make a case-by-case decision on each enumeration
+if it should be extensible or not by applying one of the two attributes. The
+language feature `ExtensibleEnums` can be thought of as implicitly adding
+`@extensible` to all enums that are not explicitly marked as `@frozen`.
+
+## Source compatibility
+
+Enabling the language feature `ExtensibleEnums` in a module that contains public
+enumerations is a source breaking change.
+Changing the annotation from `@frozen` to `@extensible` is a source breaking
+change. 
+Changing the annotation from `@extensible` to `@frozen` is a source compatible
+change and will only result in a warning code that used `@unknown default:`
+clause. This allows developers to commit to the API of an enum in a non-source
+breaking way.
+
+## Effect on ABI stability
+
+This attribute does not affect the ABI, as it is a no-op when used in a resilient library.
```

Comment:
I think this should be called out more obviously in the proposed solution section, as well as an explanation for the choice.

Reply:
Added the following section:
>In resilient modules, the `@extensible` module doesn't affect API nor ABI since
the behaviour of enumerations in modules compiled with library evolution mode
are already extensible by default. We believe that extensible enums are the
right default choice in both resilient and non-resilient modules and the new
proposed `@extensible` attribute primiarly exists to give developers a migration
path.

Let me know if you think this addresses it well enough.